tf_trees/tel.py

```diff
@@ -0,0 +1,190 @@
+    route is controlled by the smooth_step_param (see details below).
+    The trees are modeled using smooth functions and can be optimized
+    using standard continuous optimization methods (e.g., SGD).
+
+    The layer can be combined with other Keras layers and can be used anywhere
+    in the neural net.
+
+    Attributes:
+      output_logits_dim: Dimension of the output.
+      trees_num: Number of trees in the layer.
+      depth: Depth of each tree.
+      smooth_step_param: A non-negative float. Larger values make the trees
+        more likely to hard route samples (i.e., samples reach fewer leaves).
+        Note smooth_step_param = 1/gamma, where gamma is the parameter defined
+        in the TEL paper. 
+      sum_outputs: Boolean. If true, the outputs of the trees will be added,
+        leading to a 2D tensor of shape=[batch_size, output_logits_dim].
+        Otherwise, the tree outputs are not added and the layer output is
+        a 2D tensor of shape=[batch_size, trees_num * output_logits_dim].
+      parallelize_over_samples: Boolean, If true, parallelizes the updates over
+        the samples in each tree. Might lead to speedups when the number of
+        trees is small (at the cost of extra memory consumption).
+      split_initializer: A Keras initializer for the internal (splitting) nodes.
+      leaf_initializer: A Keras initializer for the leaves.
+      split_regularizer: A Keras regularizer for the internal (splitting) nodes.
+      leaf_regularizer: A Keras regularizer for the leaves.
+    Input shape: A tensor of shape=[batch_size, input_dim].
+    Output shape: A tensor of shape=[batch_size, output_logits_dim] if
+      sum_outputs=True. Otherwise, a tensor of shape=[batch_size, trees_num *
+      output_logits_dim].
+    """
+
+    def __init__(self,
+                 output_logits_dim,
+                 trees_num=1,
+                 depth=3,
+                 smooth_step_param=1.0,
+                 sum_outputs=True,
+                 parallelize_over_samples=False,
+                 split_initializer=RandomUniform(-0.01, 0.01),
+                 leaf_initializer=RandomUniform(-0.01, 0.01),
+                 split_regularizer=None,
+                 leaf_regularizer=None,
+                 **kwargs):
+      """Initializes neural trees layer."""
+      self.trees_num = trees_num
+      self.depth = depth
+      self.output_logits_dim = output_logits_dim
+      self.smooth_step_param = smooth_step_param
+      self.parallelize_over_samples = parallelize_over_samples
+      self.sum_outputs = sum_outputs
+      self.split_initializer = keras.initializers.get(split_initializer)
+      self.leaf_initializer = keras.initializers.get(leaf_initializer)
+      self.split_regularizer = keras.regularizers.get(split_regularizer)
+      self.leaf_regularizer = keras.regularizers.get(leaf_regularizer)
+      super(TEL, self).__init__(**kwargs)
+
+    def build(self, input_shape):
```

Comment:
can you add  # @Override to show that you are overriding this main method
Also a comment on what it does (e.g. preps the variables for the tree etc)